Source/WebCore/platform/graphics/cocoa/CMUtilities.mm

```diff
@@ -277,6 +282,179 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+                ret++;
+                v >>= 1;
+            }
+            return ret;
+        };
+
+        uint32_t modeBitCount = ilog(m_vorbisModeInfo->mModeCount - 1);
+        for (uint32_t thisModeBit = 0; thisModeBit < modeBitCount; ++thisModeBit)
+            m_vorbisModeMask |= 1 << thisModeBit;
+        }
+        break;
+
+    default:
+        // No need to examine the magic cookie.
+        break;
+    }
+#endif
+    m_isValid = true;
+}
+
+size_t PacketDurationParser::framesInPacket(SharedBuffer& packet)
+{
+#if !HAVE(AUDIOFORMATPROPERTY_VARIABLEPACKET_SUPPORTED)
+    UNUSED_PARAM(packet);
+    return m_constantFramesPerPacket;
+#else
+    if (m_constantFramesPerPacket)
+        return m_constantFramesPerPacket;
+
+    if (packet.isEmpty())
+        return 0;
+
+    switch (m_audioFormatID) {
+#if ENABLE(OPUS)
+    case kAudioFormatOpus:
+        {
+        constexpr uint32_t opusPacketConfigMask = 0b11111000;
+
+        UInt8 config = (*packet.data() & opusPacketConfigMask) >> 3;
+
+        return [] (uint8_t config) -> size_t {
+            // Refer to https://tools.ietf.org/html/rfc6716#section-3:
+            // Section 3.1. The TOC Byte
+            // Table 2: TOC Byte Configuration Parameters
+            // Column 4: Frame Sizes
+            // Duration in number of frames (Opus is always 48000Hz)
+            constexpr size_t frameSizeArray[] = {
+                480,
+                960,
+                1920,
+                2880,
+                480,
+                960,
+                1920,
+                2880,
+                480,
+                960,
+                1920,
+                2880,
+                480,
+                960,
+                480,
+                960,
+                120,
+                240,
+                480,
+                960,
+                120,
+                240,
+                480,
+                960,
+                120,
+                240,
+                480,
+                960,
+                120,
+                240,
+                480,
+                960,
+            };
+            if (config < std::size(frameSizeArray))
+                return frameSizeArray[config];
+
+            ASSERT_NOT_REACHED();
+            return 960; // The most common Opus frame duration.
+        }(config);
+        break;
+        }
+#endif
+#if ENABLE(VORBIS)
+    case kAudioFormatVorbis:
+        {
+        // The following calculation corresponds to the duration of the "finished audiodata"
+        // produced by the decoder from the current packet in its position within
+        // the stream, as documented by Xiph in the Vorbis I specification.
+        // It also corresponds to the delta in granule position of the packet within
+        // the same sequence of packets in an Ogg file, with the possible exception of
+        // the ultimate packet, which may be assigned a smaller delta for the purpose
+        // of trimming.
```

Comment:
Too much information below this point. Let's delete all but the first paragraph of the comment.